dbms/src/Storages/Page/V3/CheckpointFile/CPFilesWriter.cpp

```diff
@@ -89,9 +89,16 @@ bool CPFilesWriter::writeEditsAndApplyCheckpointInfo(universal::PageEntriesEdit
 
         if (rec_edit.entry.checkpoint_info.has_value())
         {
-            // for example, the s3 fullpath that was written in the previous uploaded CheckpointDataFile
-            locked_files.emplace(*rec_edit.entry.checkpoint_info.data_location.data_file_id);
-            continue;
+            const auto file_id = *rec_edit.entry.checkpoint_info.data_location.data_file_id;
+            if (!file_ids_to_compact.contains(file_id))
+            {
+                // for example, the s3 fullpath that was written in the previous uploaded CheckpointDataFile
+                locked_files.emplace(file_id);
+                continue;
+            }
+            // else we rewrite this entry data to the data file generated by this checkpoint, so that
+            // the old, fragmented data will be compacted. The outdated CheckpointDataFiles will be
+            // removed by S3GCManager after the new manifest uploaded.
```

Comment:
TODO: Add a test for whether we can override the checkpoint_info.data_location in PageDirectory